server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java

```diff
@@ -893,29 +928,40 @@ private long toMillis(int lifespan) {
       }
    }
 
-   protected Object writeResponse(Channel ch, Object response) {
-      try {
-         if (response != null) {
-            if (log.isTraceEnabled()) log.tracef("Write response %s", response);
-            if (response instanceof ByteBuf[]) {
-               for (ByteBuf buf : (ByteBuf[]) response) {
-                  ch.write(buf, ch.voidPromise());
-                  ch.flush();
-               }
-            } else if (response instanceof byte[]) {
-               ch.writeAndFlush(wrappedBuffer((byte[]) response), ch.voidPromise());
-            } else if (response instanceof CharSequence) {
-               ch.writeAndFlush(Unpooled.copiedBuffer((CharSequence) response, CharsetUtil.UTF_8), ch.voidPromise());
-            } else if (response instanceof PartialResponse) {
-               return response;
-            } else {
-               ch.writeAndFlush(response, ch.voidPromise());
+   private void writeResponseOrThrowable(Channel ch, Object response, Throwable throwable) {
+      if (throwable != null) {
+         Throwable cause = CompletableFutures.extractException(throwable);
+         if (cause instanceof IOException || cause instanceof NumberFormatException) {
+            cause = wrapBadFormat(cause);
+         } else {
+            cause = wrapServerError(cause);
+         }
+         ch.pipeline().fireExceptionCaught(cause);
+      } else {
+         writeResponse(ch, response);
+      }
+   }
+
+   private Object writeResponse(Channel ch, Object response) {
+      if (response != null) {
+         if (log.isTraceEnabled())
+            log.tracef("Write response using instance @%x: %s", System.identityHashCode(this), response);
+         if (response instanceof ByteBuf[]) {
+            for (ByteBuf buf : (ByteBuf[]) response) {
+               ch.write(buf, ch.voidPromise());
+               ch.flush();
             }
+         } else if (response instanceof byte[]) {
+            ch.writeAndFlush(wrappedBuffer((byte[]) response), ch.voidPromise());
+         } else if (response instanceof CharSequence) {
+            ch.writeAndFlush(Unpooled.copiedBuffer((CharSequence) response, CharsetUtil.UTF_8), ch.voidPromise());
+         } else if (response instanceof PartialResponse) {
+            return response;
+         } else {
+            ch.writeAndFlush(response, ch.voidPromise());
          }
-         return null;
-      } finally {
-         resetParams();
       }
+      return null;
    }
 
    Object createStatsResponse() {
```

Comment:
I totally forgot about `getStats` but, luckily, it does not block since it only reports the local stats :)

Reply:
Can we configure a store with this cache though?

Reply:
Right! `stats.getCurrentNumberOfEntries()` may block in the cache store :(
Should I replace it with `getApproximateEntries()` to avoid complicating the code?

Reply:
Sure, sounds fine to me. I doubt anyone is going to be using this much anyways as they can just use our stats endpoint :)